example/list.sw

```diff
@@ -50,265 +46,250 @@ vvvvvvvvvvvv       vvvvvvvvvvvvvvvvvvvvvvvvv     vvv
 ^^^^^^^^^^^^       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   head                         tail
 */
-
 /*******************************************************************/
 /*                         ARITH PRIMITIVES                        */
 /*******************************************************************/
-
 // bitlength of a number (shifting by one)
-def naive_len : Int -> Int = \i.
-  if (i == 0) {0} {1 + naive_len (i/2)}
+def naive_len: Int -> Int -> Int 
+  = \acc. \i.
+  if (i == 0) {acc} {naive_len (1 + acc) (i / 2)}
 end
 
+
 // modulus function (%)
-def mod : Int -> Int -> Int = \i.\m.
-  i - m * (i / m)
-end
+def mod: Int -> Int -> Int = \i. \m. i - m * (i / m) end
+
 
 // f X Y = Y * 2^(-X)
-def shiftL : Int -> Int -> Int = \s.\i.
-  i / 2^s
-end
+def shiftL: Int -> Int -> Int = \s. \i. i / 2 ^ s end
+
 
 // f X Y = Y * 2^(X)
-def shiftR : Int -> Int -> Int = \s.\i.
-  i * 2^s
-end
+def shiftR: Int -> Int -> Int = \s. \i. i * 2 ^ s end
+
 
 // shift by (8bit) bytes
-def shiftLH : Int -> Int -> Int = \s. shiftL (s*8) end
-def shiftRH : Int -> Int -> Int = \s. shiftR (s*8) end
+def shiftLH: Int -> Int -> Int = \s. shiftL (s * 8) end
 
-// bitlength of a number (shifting by 64)
-def len : Int -> Int = \i.
-  let next = i / 2^64 in
-  if (next == 0) {naive_len i} {64 + len next}
+def shiftRH: Int -> Int -> Int = \s. shiftR (s * 8) end
+
+
+// bitlength of a number using an accumulator (shifting by 64)
+def len_accum: Int -> Int -> Int 
+  = \acc. \i.
+  let next = i / 2 ^ 64 in
+  if (next == 0) {naive_len acc i} {len_accum (acc + 64) next}
 end
 
+
+// bitlength of a number (uses an accumulator and shifts by 64 bits)
+def len: Int -> Int = len_accum 0 end
+
+
 /*******************************************************************/
 /*                       helper functions                          */
 /*******************************************************************/
+def getLenPart: Int -> Int = \i. mod (i / 2) (2 ^ 7) end
+
+def setLenPart: Int -> Int = \i. 2 * mod i (2 ^ 7) end
 
-def getLenPart : Int -> Int = \i. mod (i/2) (2^7) end
-def setLenPart : Int -> Int = \i. 2 * (mod i (2^7)) end
 
 // Split length into 7-bit parts and prefix all but last with 1
-def to1numA : Int -> Int * Int = \i.
+def to1numA: Int -> (Int * Int) 
+  = \i.
   let nextPart = shiftL 7 i in
-  if (nextPart == 0)
-    { ((2 * i), 1) } /* last part */
-    { let p = to1numA nextPart in
-      ( 1 /* header isEnd bit */ + setLenPart i + shiftRH 1 (fst p)
-      , 1 + snd p
-      )
-    }
+  if (nextPart == 0) {(2 * i, 1 /* last part */)} {
```

Comment:
Hmm, the reformatting moved the `last part` comment inside some parens...

Reply:
I moved the comment before the expression, but the formatted inserts an empty line before it.
I guess that is better, but will needs some tweaks if we want to use it for all files.

Reply:
Thinking about this more, I am actually surprised that it does not print `{(2 * i, 1)} /* last part */`.  I thought it was supposed to do that; I will have to think about whether this is a bug.

Reply:
Yeah, getting auto-formatted comments to look good in all situations is surprisingly tough.  But I would be very happy to collect more examples of where it breaks down to see if we can continue tweaking the output.